test/Test/Pos/Types/SlottingSpec.hs

```diff
@@ -4,26 +4,72 @@ module Test.Pos.Types.SlottingSpec
 import           Universum
 
-import           Pos.Arbitrary.Core    ()
-import           Pos.Types             (SlotId, flattenSlotId, unflattenSlotId)
+import           Test.Hspec            (Expectation, Spec, anyErrorCall, describe)
+import           Test.Hspec.QuickCheck (prop)
+import           Test.QuickCheck       (NonNegative (..), Positive (..), Property, (===),
+                                        (==>))
+
+import           Pos.Arbitrary.Core    (EoSToIntOverflow (..))
+import           Pos.Types             (EpochIndex (..), EpochOrSlot (..), SlotId,
+                                        flattenSlotId, unflattenSlotId)
+import           Test.Pos.Util         (shouldThrowException, (.=.))
 
 spec :: Spec
-spec = describe "SlotId" $ do
-    describe "Ord" $ do
-        prop "is consistent with flatten/unflatten"
-            flattenOrdConsistency
+spec = describe "Slotting" $ do
+    describe "SlotId" $ do
+        describe "Ord" $ do
+            prop "is consistent with flatten/unflatten"
+                flattenOrdConsistency
 
-    describe "flattening" $ do
-        prop "unflattening after flattening returns original SlotId"
-            flattenThenUnflatten
+        describe "flattening" $ do
+            prop "unflattening after flattening returns original SlotId"
+                flattenThenUnflatten
+
+    describe "EpochOrSlot" $ do
+        prop "succ . pred = id" predThenSucc
+        prop "Using 'pred' with 'minBound :: EpochOrSlot' triggers an exception"
+            predToMinBound
+        prop "pred . succ = id" succThenPred
+        prop "Using 'succ' with 'maxBound :: EpochOrSlot' triggers an exception"
+            succToMaxBound
+        prop "from . toEnum = id @Int" toFromEnum
+        prop "toEnum . fromEnum = id @EpochOrSlot" fromToEnum
+        prop "calling 'fromEnum' with a result greater than 'maxBound :: Int' will raise \
+              \ an exception" fromEnumOverflow
+        prop "calling 'toEnum' with a negative number will raise an exception"
+            toEnumNegative
 
 flattenOrdConsistency :: SlotId -> SlotId -> Property
 flattenOrdConsistency a b = a `compare` b === flattenSlotId a `compare` flattenSlotId b
 
 flattenThenUnflatten :: SlotId -> Property
 flattenThenUnflatten si = si === unflattenSlotId (flattenSlotId si)
+
+predThenSucc :: EpochOrSlot -> Property
+predThenSucc eos = eos > minBound ==> succ (pred eos) === eos
+
+predToMinBound :: Expectation
+predToMinBound =
+    shouldThrowException pred anyErrorCall (EpochOrSlot $ Left (EpochIndex 0))
+
+succThenPred :: EpochOrSlot -> Property
+succThenPred eos = eos < maxBound ==> pred (succ eos) === eos
+
+succToMaxBound :: Expectation
+succToMaxBound = shouldThrowException succ anyErrorCall (maxBound :: EpochOrSlot)
+
+toFromEnum :: NonNegative Int -> Property
+toFromEnum (getNonNegative -> int) = fromEnum (toEnum @EpochOrSlot int) === int
+
+fromToEnum :: EpochOrSlot -> Property
+fromToEnum = toEnum . fromEnum .=. identity
```

Comment:
Hm, it actually doesn't always hold because if `EpochOrSlot` is e. g. `maxBound` it will throw an exception. I guess the reason it works now is that in `instance Arbitrary EpochIndex` we always generate reasonable epochs (at most 5 * 10¹²).  Maybe it makes sense to handle this case separately (just to make sure this test won't break if we increase `maxReasonableEpoch` or change `Arbitrary` instance), but I don't insist.